java/javax/el/ImportHandler.java

```
@@ -0,0 +1,491 @@
+
+        if (lastPeriod < 0) {
+            throw new ELException(Util.message(
+                    null, "importHandler.invalidStaticName", name));
+        }
+
+        String className = name.substring(0, lastPeriod);
+        String fieldOrMethodName = name.substring(lastPeriod + 1);
+
+        Class<?> clazz = findClass(className, true);
+
+        if (clazz == null) {
+            throw new ELException(Util.message(
+                    null, "importHandler.invalidClassNameForStatic",
+                    className, name));
+        }
+
+        boolean found = false;
+
+        for (Field field : clazz.getFields()) {
+            if (field.getName().equals(fieldOrMethodName)) {
+                int modifiers = field.getModifiers();
+                if (Modifier.isStatic(modifiers) &&
+                        Modifier.isPublic(modifiers)) {
+                    found = true;
+                    break;
+                }
+            }
+        }
+
+        if (!found) {
+            for (Method method : clazz.getMethods()) {
+                if (method.getName().equals(fieldOrMethodName)) {
+                    int modifiers = method.getModifiers();
+                    if (Modifier.isStatic(modifiers) &&
+                            Modifier.isPublic(modifiers)) {
+                        found = true;
+                        break;
+                    }
+                }
+            }
+        }
+
+        if (!found) {
+            throw new ELException(Util.message(null,
+                    "importHandler.staticNotFound", fieldOrMethodName,
+                    className, name));
+        }
+
+        Class<?> conflict = statics.get(fieldOrMethodName);
+        if (conflict != null) {
+            throw new ELException(Util.message(null,
+                    "importHandler.ambiguousStaticImport", name,
+                    conflict.getName() + '.' +  fieldOrMethodName));
+        }
+
+        statics.put(fieldOrMethodName, clazz);
```

Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **333, 334, and 340**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 If the key already exists, the code should throw an error at line: **335**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.